### PR TITLE
Add pgo instrument job

### PR DESCRIPTION
**Pull request by agocke**
Adds a new flag for building packages instrumented with PGO data. The
packages have new names as well, <package>.PGO.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
Just to confirm I have the right mindset for reviewing this, the idea here is to produce a parallel set of packages with .PGO appended to the package name, flow these to the dotnet-optimization tooling which will pull these down and run training. A pipline for that repo will then create and upload the new PGO/IBC result packages, which dotnet/runtime will then consume.

Is that correct?

**Comment by agocke** [comment]
Basically, yes. I made some assumptions, like the best way to segment the build was creating ".PGO" packages instead of something else. If you think we should do something else, like create a different version, happy to do that instead.

**Comment by jkoritzinsky** [comment]
I think that should work. We just need to make sure that going down this path won't make it difficult for us to also generate IBC/managed PGO info from a non-PGO instrumented runtime.

File: eng/pipelines/runtime.yml [hunk 1]
```diff
@@ -722,6 +722,20 @@ jobs:
       liveRuntimeBuildConfig: release
       liveLibrariesBuildConfig: ${{ variables.debugOnPrReleaseOnRolling }}
 
+  #
+  # PGO Build
+  #
```
Comment (jkoritzinsky) [inline eng/pipelines/runtime.yml:727]: Don't forget to add the coreclr PGO build as well here so we have an instrumented runtime.

File: eng/pipelines/installer/jobs/base-job.yml [hunk 2]
```diff
@@ -38,8 +38,8 @@ parameters:
   platforms: []
 
 jobs:
-- job: ${{ format('installer_{0}_{1}_{2}_{3}', parameters.runtimeFlavor, parameters.runtimeVariant, coalesce(parameters.name, parameters.platform), parameters.buildConfig) }}
-  displayName: ${{ format('Installer Build and Test {0} {1} {2} {3}', parameters.runtimeFlavor, parameters.runtimeVariant, coalesce(parameters.name, parameters.platform), parameters.buildConfig) }}
+- job: ${{ format('installer_{0}_{1}_{2}_{3}_{4}', parameters.pgoInstrument, parameters.runtimeFlavor, parameters.runtimeVariant, coalesce(parameters.name, parameters.platform), parameters.buildConfig) }}
```
Comment (jkoritzinsky) [inline eng/pipelines/installer/jobs/base-job.yml:41]: You'll need to update the `dependsOn:` section as well as the intermediate artifact download to pull down the PGO instrumented CoreCLR build to package.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
   <PropertyGroup>
     <SharedFrameworkName>Microsoft.NETCore.App</SharedFrameworkName>
+    <SharedFrameworkName Condition="'$(PgoOptimize)' != ''">$(SharedFrameworkName).PGO</SharedFrameworkName>
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:6]: You should also mark the PGO package variants as non-Shipping.

You'll likely also need to include the `pgort*.dll` file in the platform manifest for the PGO-instrumented build.

**Review by jkoritzinsky — COMMENTED** [review]
Looks good so far, but there's still more work to be done to wire up packaging correctly.

**Comment by agocke** [comment]
@jkoritzinsky I'd like to change the job name to include the words "PGO" when 'parameter.pgoInstrument' is true, but I can't seem to find any conditional expression syntax. Suggestions?

**Comment by jkoritzinsky** [comment]
I think the only way to do conditional statements is with `${{ if cond }}`. I don't think there's another way.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props [hunk 4]
```diff
@@ -3,7 +3,8 @@
 
   <PropertyGroup>
     <SharedFrameworkName>Microsoft.NETCore.App</SharedFrameworkName>
-    <SharedFrameworkName Condition="'$(PgoOptimize)' != ''">$(SharedFrameworkName).PGO</SharedFrameworkName>
+    <SharedFrameworkName Condition="'$(PgoInstrument)' != ''">$(SharedFrameworkName).PGO</SharedFrameworkName>
+    <IsShipping Condition="'$(PgoInstrument)' != ''">true</IsShipping>
```
Comment (jkoritzinsky) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:7]: We shouldn't mark the package as shipping if it's the PGO-instrumented one.

```suggestion
    <IsShipping Condition="'$(PgoInstrument)' != ''">false</IsShipping>
```

**Comment by agocke** [comment]
@jkoritzinsky Hmm, it looks like the name Microsoft.NETCore.App is hardcoded in KnownFrameworkReference -- is there a good way to rewrite this to use a different name? Otherwise it seems like this may be complicated, compared to using a special version number

**Comment by jkoritzinsky** [comment]
I think the easiest solution in this case is to just skip tests for the PGO installer leg. We're already going to be running the tests on the non-instrumented leg, so we don't need to run them on this leg.

**Comment by agocke** [comment]
@jkoritzinsky I think this is ready for re-review. Should I move these new definitions into the runtime-official job for asset publishing? Or leave them here?

**Comment by jkoritzinsky** [comment]
I think it would be worth leaving it here for continued validation that we don't break it, but we definitely need to add it to the official build pipeline as well for asset publishing.

**Review by jkoritzinsky — COMMENTED** [review]
LGTM once it's added to the official build pipeline.

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**